app/src/main/java/com/sdp/movemeet/view/chat/ChatActivity.java

```diff
@@ -343,28 +343,30 @@ protected void onActivityResult(int requestCode, int resultCode, Intent data) {
             if (resultCode == RESULT_OK && data != null) {
                 final Uri uri = data.getData();
                 Log.d(TAG, "Uri: " + uri.toString());
-                Message tempMessage = new Message(fullNameString, null, userId, LOADING_IMAGE_URL);
-                chatRoom.push().setValue(tempMessage, new DatabaseReference.CompletionListener() {
-                    @Override
-                    public void onComplete(DatabaseError databaseError, DatabaseReference databaseReference) {
-                        if (databaseError != null) {
-                            Log.w(TAG, "Unable to write message to database.", databaseError.toException());
-                            return;
-                        }
-                        // Building a StorageReference and then uploading the image file
-                        String key = databaseReference.getKey();
-                        StorageReference fileRef = storageReference
-                                .child(CHATS_CHILD)
-                                .child(CHAT_ROOM_ID)
-                                .child(key)
-                                .child(uri.getLastPathSegment());
-                        putImageInStorage(fileRef, uri, key);
-                    }
-                });
+                createTempMessage(uri, fullNameString, userId);
             }
         }
     }
 
+    @VisibleForTesting(otherwise=VisibleForTesting.PRIVATE) // making this method always public for testing and private otherwise
+    public void createTempMessage(Uri uri, String fullNameString, String userId) {
+        Message tempMessage = new Message(fullNameString, "Image loading...", userId, LOADING_IMAGE_URL, Long.toString(new Date().getTime()));
+        chatRoom.push().setValue(tempMessage, new DatabaseReference.CompletionListener() {
+            @Override
+            public void onComplete(DatabaseError databaseError, DatabaseReference databaseReference) {
+                if (databaseError != null) {
+                    Log.w(TAG, "Unable to write message to database.", databaseError.toException());
+                    return;
+                }
+                // Building a StorageReference and then uploading the image file
+                String key = databaseReference.getKey();
+                StorageReference fileRef = storageReference.child(CHATS_CHILD).child(CHAT_ROOM_ID).child(key).child(uri.getLastPathSegment());
```

Comment:
We don't have a Storage abstraction? (genuine question, I thought someone had made one)

Reply:
No, indeed. This is something I initially planned to do during this week in addition to finishing implementing Firebase Realtime Database in the ChatActivity and fixing other chat related bugs, but I finally didn't have the time to do it.